File: agenkit/composition/parallel.py
"""Parallel agent composition pattern."""

import asyncio
from dataclasses import dataclass
from typing import List, Optional

from agenkit.interfaces import Agent, Message


@dataclass
class AgentResult:
    """Result from a single agent execution."""

    agent_name: str
    message: Optional[Message]
    error: Optional[Exception]


class ParallelAgent(Agent):
    """Agent that executes multiple agents concurrently and combines their results."""

    def __init__(self, name: str, agents: List[Agent]):
        """Initialize parallel agent.

        Args:
            name: Name of this parallel agent
            agents: List of agents to execute in parallel

        Raises:
            ValueError: If agents list is empty
        """
        if not agents:
            raise ValueError("Parallel agent requires at least one agent")

        self._name = name
        self._agents = agents

    @property
    def name(self) -> str:
        """Return the name of the parallel agent."""
        return self._name

    @property
    def capabilities(self) -> list[str]:
        """Return combined capabilities of all agents."""
        caps_set = set()
        for agent in self._agents:
            caps_set.update(agent.capabilities)
        caps = list(caps_set)
        caps.append("parallel")
        return caps

    async def process(self, message: Message) -> Message:
        """Execute all agents in parallel and combine their results.

        Args:
            message: Input message

        Returns:
            Combined message with results from all agents

        Raises:
            Exception: If any agent fails
        """
        # Create tasks for all agents
        tasks = [self._execute_agent(agent, message) for agent in self._agents]

        # Wait for all agents to complete
        results = await asyncio.gather(*tasks)

        # Check for errors
        errors = []
        for result in results:
            if result.error:
                errors.append(f"{result.agent_name}: {result.error}")

        if errors:
            raise Exception(f"Parallel execution had errors: {'; '.join(errors)}")

        # Combine all responses
        return self._combine_responses(results)

    async def _execute_agent(self, agent: Agent, message: Message) -> AgentResult:
        """Execute a single agent and return the result.

        Args:
            agent: Agent to execute
            message: Input message

        Returns:
            Agent result containing message or error
        """
        try:
            result = await agent.process(message)
            return AgentResult(agent_name=agent.name, message=result, error=None)
        except Exception as e:
            return AgentResult(agent_name=agent.name, message=None, error=e)

    def _combine_responses(self, results: List[AgentResult]) -> Message:
        """Combine multiple agent responses into a single message.

        Args:
            results: List of agent results

        Returns:
            Combined message with all agent outputs
        """
        content_parts = []
        combined_metadata = {}

        for result in results:
            if result.message:
                content_parts.append(f"[{result.agent_name}]: {result.message.content}")

                # Merge metadata with agent name prefix
                for key, value in result.message.metadata.items():
                    prefixed_key = f"{result.agent_name}.{key}"
                    combined_metadata[prefixed_key] = value

        response = Message(role="agent", content="\n".join(content_parts))
        response.metadata = combined_metadata
        return response

    def get_agents(self) -> List[Agent]:
        """Return the list of agents that run in parallel."""
        return self._agents

File: agenkit-cpp/src/composition/parallel.cpp
/**
 * @file parallel.cpp
 * @brief Implementation of ParallelAgent
 */

#include "agenkit/composition/parallel.hpp"
#include <stdexcept>
#include <set>
#include <sstream>

namespace agenkit {
namespace composition {

using namespace core;

ParallelAgent::ParallelAgent(
    std::string name,
    std::vector<std::shared_ptr<Agent>> agents
) : name_(std::move(name)), agents_(std::move(agents)) {
    if (agents_.empty()) {
        throw std::invalid_argument("Parallel agent requires at least one agent");
    }
}

std::vector<std::string> ParallelAgent::capabilities() const {
    std::set<std::string> cap_set;

    for (const auto& agent : agents_) {
        for (const auto& cap : agent->capabilities()) {
            cap_set.insert(cap);
        }
    }

    std::vector<std::string> capabilities(cap_set.begin(), cap_set.end());
    capabilities.push_back("parallel");

    return capabilities;
}

std::future<Result<Message, AgentError>>
ParallelAgent::process(Message message) {
    return std::async(std::launch::async, [this, message]() {
        // Launch all agents concurrently
        std::vector<std::future<Result<Message, AgentError>>> futures;
        futures.reserve(agents_.size());

        for (const auto& agent : agents_) {
            futures.push_back(agent->process(message));
        }

        // Collect results
        std::vector<std::string> content_parts;
        std::vector<std::string> errors;

        for (size_t i = 0; i < futures.size(); ++i) {
            try {
                auto result = futures[i].get();

                if (result.is_err()) {
                    errors.push_back(agents_[i]->name() + ": " + result.unwrap_err().message());
                } else {
                    content_parts.push_back(
                        "[" + agents_[i]->name() + "]: " + result.unwrap().content_as_str()
                    );
                }
            } catch (const std::exception& e) {
                errors.push_back(agents_[i]->name() + ": " + std::string(e.what()));
            }
        }

        if (!errors.empty()) {
            std::ostringstream oss;
            oss << "Parallel execution had errors: ";
            for (size_t i = 0; i < errors.size(); ++i) {
                if (i > 0) oss << "; ";
                oss << errors[i];
            }
            return Result<Message, AgentError>::err(AgentError(
                AgentErrorType::ProcessingError,
                oss.str()
            ));
        }

        // Combine content
        std::ostringstream combined;
        for (size_t i = 0; i < content_parts.size(); ++i) {
            if (i > 0) combined << "\n";
            combined << content_parts[i];
        }

        return Result<Message, AgentError>::ok(Message::with_text("agent", combined.str()));
    });
}

} // namespace composition
} // namespace agenkit
